src/main/java/com/digitalpebble/storm/crawler/bolt/SiteMapParserBolt.java

```diff
@@ -0,0 +1,178 @@
+ * limitations under the License.
+ */
+
+package com.digitalpebble.storm.crawler.bolt;
+
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.LoggerFactory;
+
+import backtype.storm.task.OutputCollector;
+import backtype.storm.task.TopologyContext;
+import backtype.storm.topology.OutputFieldsDeclarer;
+import backtype.storm.topology.base.BaseRichBolt;
+import backtype.storm.tuple.Fields;
+import backtype.storm.tuple.Tuple;
+import backtype.storm.tuple.Values;
+
+import com.digitalpebble.storm.crawler.Constants;
+import com.digitalpebble.storm.crawler.persistence.Status;
+import com.digitalpebble.storm.crawler.protocol.HttpHeaders;
+import com.digitalpebble.storm.crawler.util.KeyValues;
+
+import crawlercommons.sitemaps.AbstractSiteMap;
+import crawlercommons.sitemaps.SiteMap;
+import crawlercommons.sitemaps.SiteMapIndex;
+import crawlercommons.sitemaps.SiteMapURL;
+import crawlercommons.sitemaps.SiteMapURL.ChangeFrequency;
+
+/**
+ * Extracts URLs from sitemap files. The parsing is triggered by the presence of
+ * 'isSitemap=true' in the metadata. Any tuple which does not have this
+ * key/value in the metadata is simply passed on to the default stream, whereas
+ * any URLs extracted from the sitemaps is sent to the 'status' field.
+ **/
+public class SiteMapParserBolt extends BaseRichBolt {
+
+    private OutputCollector collector;
+    public static final String isSitemapKey = "isSitemap";
+
+    private boolean strictMode = false;
+
+    private static final org.slf4j.Logger LOG = LoggerFactory
+            .getLogger(SiteMapParserBolt.class);
+
+    @Override
+    public void execute(Tuple tuple) {
+        HashMap<String, String[]> metadata = (HashMap<String, String[]>) tuple
+                .getValueByField("metadata");
+
+        // TODO check that we have the right number of fields ?
+        String isBoolean = KeyValues.getValue(isSitemapKey, metadata);
+        if (StringUtils.isBlank(isBoolean)
+                || !isBoolean.equalsIgnoreCase("true")) {
+            // just pass it on
+            this.collector.emit(tuple.getValues());
+            this.collector.ack(tuple);
+            return;
+        }
+
+        // it does have the right key/value
+        byte[] content = tuple.getBinaryByField("content");
+        String url = tuple.getStringByField("url");
+        String ct = KeyValues.getValue(HttpHeaders.CONTENT_TYPE, metadata);
+        List<Values> outlinks = parseSiteMap(url, content, ct);
+        if (outlinks == null) {
+            // likely to happen ad lib - just ack
+            // error has been logged
+            this.collector.ack(tuple);
+            return;
+        }
+
+        // send to status stream
+        for (Values ol : outlinks) {
+            collector.emit(Constants.StatusStreamName, new Values(ol));
```

Comment:
Isn't this odd? `new Values(ol)` doesn't create a copy of the Values. Shouldn't it be `collector.emit(Constants.StatusStreamName, ol);`